File: backend/controllers/authController.js
const bcrypt = require('bcryptjs');
const jwt = require('jsonwebtoken');
const validator = require('validator');
const { USER_PROFILE_TYPES, JWT_SECRET_KEY, MIN_PASSWORD_LENGTH } = require('../config/constants');

// --- Configuration ---
// const JWT_SECRET = process.env.JWT_SECRET || 'your-super-secret-and-long-key'; // Now from constants.js
const BCRYPT_SALT_ROUNDS = 10; // This can also be moved to constants.js if used elsewhere
// const MIN_PASSWORD_LENGTH = 8; // Now from constants.js

// --- Mock Database ---
// In a real application, this would interact with a PostgreSQL database.
// For now, we'll use an in-memory array to store users.
const mockUsersDb = [];
let userIdCounter = 1;

// --- Controller Functions ---

/**
 * Registers a new user.
 * POST /auth/register
 * Request body: { email, password, profile_type }
 */
exports.register = async (req, res) => {
    try {
        const { email, password, profile_type } = req.body;

        // 1. Validate input
        if (!email || !password || !profile_type) {
            return res.status(400).json({ message: 'Email, password, and profile_type are required.' });
        }

        if (!validator.isEmail(email)) {
            return res.status(400).json({ message: 'Invalid email format.' });
        }

        if (password.length < MIN_PASSWORD_LENGTH) {
            return res.status(400).json({ message: `Password must be at least ${MIN_PASSWORD_LENGTH} characters long.` });
        }

        // const allowedProfileTypes = ['freelance', 'accountant', 'sme', 'large_enterprise']; // Now USER_PROFILE_TYPES from constants
        if (!USER_PROFILE_TYPES.includes(profile_type)) {
            return res.status(400).json({ message: `Invalid profile_type. Must be one of: ${USER_PROFILE_TYPES.join(', ')}` });
        }

        // 2. Check if email already exists (mocked)
        if (mockUsersDb.find(user => user.email === email)) {
            return res.status(409).json({ message: 'Email already exists.' });
        }

        // 3. Hash the password
        const passwordHash = await bcrypt.hash(password, BCRYPT_SALT_ROUNDS);

        // 4. Create new user (mocked)
        const newUser = {
            id: userIdCounter++,
            email,
            passwordHash, // In a real db, this would be 'password_hash'
            profileType: profile_type, // In a real db, this would be 'profile_type'
            createdAt: new Date(), // In a real db, this would be 'created_at'
            updatedAt: new Date()  // In a real db, this would be 'updated_at'
        };
        mockUsersDb.push(newUser);

        // 5. Return success response (excluding password)
        // In a real app, you might return the full user object or just a success message.
        const userToReturn = { ...newUser };
        delete userToReturn.passwordHash; // Never return the password hash

        res.status(201).json({ message: 'User registered successfully.', user: userToReturn });

    } catch (error) {
        console.error('Registration error:', error);
        res.status(500).json({ message: 'Internal server error during registration.' });
    }
};

/**
 * Logs in an existing user.
 * POST /auth/login
 * Request body: { email, password }
 */
exports.login = async (req, res) => {
    try {
        const { email, password } = req.body;

        // 1. Validate input
        if (!email || !password) {
            return res.status(400).json({ message: 'Email and password are required.' });
        }

        if (!validator.isEmail(email)) {
            return res.status(400).json({ message: 'Invalid email format.' });
        }

        // 2. Find user by email (mocked)
        const user = mockUsersDb.find(u => u.email === email);
        if (!user) {
            return res.status(401).json({ message: 'Invalid email or password.' }); // Generic message for security
        }

        // 3. Compare password with stored hash
        const isMatch = await bcrypt.compare(password, user.passwordHash);
        if (!isMatch) {
            return res.status(401).json({ message: 'Invalid email or password.' }); // Generic message
        }

        // 4. Generate JWT
        const tokenPayload = {
            userId: user.id,
            profileType: user.profileType // Using profileType as stored in mock DB
        };
        const token = jwt.sign(tokenPayload, JWT_SECRET_KEY, { expiresIn: '1h' }); // Token expires in 1 hour

        // 5. Return JWT
        res.status(200).json({ message: 'Login successful.', token });

    } catch (error) {
        console.error('Login error:', error);
        res.status(500).json({ message: 'Internal server error during login.' });
    }
};

File: backend/controllers/auth_controller.py
# Conceptual Implementation for User Authentication API Endpoints

# Password Hashing:
# We will use bcrypt for password hashing.
# Library: bcrypt (e.g., `pip install bcrypt`)

# --- API Endpoints ---

# 1. User Registration
# Endpoint: POST /api/auth/register
# Description: Registers a new user.
# Request Payload (JSON):
# {
#   "email": "user@example.com",
#   "password": "securepassword123",
#   "profile_type": "freelance"  # or 'accountant', 'sme', 'large_enterprise'
# }
# Response Payload (Success - 201 Created):
# {
#   "user_id": 1,
#   "email": "user@example.com",
#   "profile_type": "freelance",
#   "message": "User registered successfully."
# }
# Response Payload (Error - 400 Bad Request, e.g., email already exists, invalid input):
# {
#   "error": "Email already exists or invalid data provided."
# }
# Conceptual Steps:
# 1. Validate request payload (email format, password strength, valid profile_type).
# 2. Check if email already exists in the database.
# 3. Hash the password using bcrypt.
# 4. Create a new user record in the database with the hashed password and other details.
# 5. Return the newly created user's information (excluding password) or an access token.

# 2. User Login
# Endpoint: POST /api/auth/login
# Description: Logs in an existing user.
# Request Payload (JSON):
# {
#   "email": "user@example.com",
#   "password": "securepassword123"
# }
# Response Payload (Success - 200 OK):
# {
#   "token": "your_jwt_access_token",  # Or session token
#   "user_id": 1,
#   "email": "user@example.com",
#   "profile_type": "freelance",
#   "message": "Login successful."
# }
# Response Payload (Error - 401 Unauthorized, e.g., invalid credentials):
# {
#   "error": "Invalid email or password."
# }
# Response Payload (Error - 404 Not Found, e.g., user does not exist):
# {
#   "error": "User not found."
# }
# Conceptual Steps:
# 1. Validate request payload.
# 2. Find the user by email in the database.
# 3. If user exists, compare the provided password with the stored hashed password using bcrypt.
# 4. If passwords match, generate a JWT (JSON Web Token) or session token.
# 5. Return the token and user information (excluding password).

# --- Placeholder Functions (to be implemented with a framework) ---

def register_user(request_data):
    """
    Conceptual function to handle user registration.
    - Validate data
    - Check for existing user
    - Hash password
    - Store user in DB
    - Return response
    """
    email = request_data.get("email")
    password = request_data.get("password")
    profile_type = request_data.get("profile_type")

    # --- Pseudocode ---
    # if not is_valid_email(email) or not is_strong_password(password) or not is_valid_profile_type(profile_type):
    #     return {"error": "Invalid data provided."}, 400
    #
    # if user_exists_in_db(email):
    #     return {"error": "Email already exists."}, 400
    #
    # hashed_password = hash_with_bcrypt(password)
    # new_user = create_user_in_db(email, hashed_password, profile_type)
    #
    # return {
    #     "user_id": new_user.id,
    #     "email": new_user.email,
    #     "profile_type": new_user.profile_type,
    #     "message": "User registered successfully."
    # }, 201
    pass

def login_user(request_data):
    """
    Conceptual function to handle user login.
    - Validate data
    - Find user
    - Verify password
    - Generate token
    - Return response
    """
    email = request_data.get("email")
    password = request_data.get("password")

    # --- Pseudocode ---
    # user = find_user_in_db(email)
    # if not user:
    #     return {"error": "User not found."}, 404
    #
    # if verify_password_with_bcrypt(password, user.hashed_password):
    #     token = generate_jwt_token(user.id, user.email)
    #     return {
    #         "token": token,
    #         "user_id": user.id,
    #         "email": user.email,
    #         "profile_type": user.profile_type,
    #         "message": "Login successful."
    #     }, 200
    # else:
    #     return {"error": "Invalid email or password."}, 401
    pass
